Curie keybindings

ChiragGhanshani opened the pull request:
Fixes #31.

I have created keybindings for when the right and left arrow keys are pressed. The bindings for scrolling are still pending implementation.

I haven't added the actions that are triggered when the key presses are detected yet because I'm having trouble getting a reference to the buttons.
- [ ] Test the left and right keys

ChiragGhanshani committed:
Added several keybindings

- Right and left arrow keys can be used for horizontal scrolling.
- Up and down arrow keys can be used for vertical scrolling.
- Page down and page up keys can be used to go to the next page and the
  previous page.

ChiragGhanshani commented:
@zmughal @StayerX I'll probably need help with this one.

zmughal commented:
I'm going to rebase this off of master now, since there is an extra commit that doesn't belong.

zmughal commented:
For the testing, look at [this code](https://mail.gnome.org/archives/gtkmm-list/2007-June/msg00115.html) that uses `gdk_event_put`. It will allow for simulating key presses.

zmughal commented on lib/Renard/Curie/App.pm:
```diff
@@ -59,6 +59,15 @@ sub key_pressed {
 		$self->page_document_component->set_current_page_forward;
 	}
+	elsif($event->keyval == Gtk3::Gdk::KEY_Left){
+		$self->page_document_component->set_current_page_back;
```
Just an FYI, we need to use PgDown and PgUp for changing the page. Left and right arrows will be useful when we have zoom and need horizontal scrolling.

zmughal commented on lib/Renard/Curie/Component/PageDrawingArea.pm:
```diff
@@ -148,6 +164,8 @@ sub can_move_to_next_page {
 	$self->current_page_number < $self->document->last_page_number;
 }
 
+has scrolled_window => (is => 'rw');
```
You should move this attribute up to the top with the other attributes.

Also, this attribute will still be undefined until you set it. I recommend doing that in setup_drawing_area:

``` perl
$self->scrolled_window(  $scrolled_window );
```

zmughal commented:
Looking good so far! I just ran the code too.

ChiragGhanshani commented:
@zmughal the video in the skype call is about making tests for the gui right?

zmughal commented:
yes, how to run them locally

ChiragGhanshani commented:
@zmughal the tests are all that remain for issue #31

zmughal commented:
In addition to the link above in the comment https://github.com/project-renard/curie/pull/52#issuecomment-201832175, check out this code in the test for Gtk3-Perl https://github.com/GNOME/perl-Gtk3/blob/master/t/zz-GdkEvent.t#L425-L439

ChiragGhanshani commented:
what about using https://developer.gnome.org/gdk3/stable/gdk3-Testing.html#gdk-test-simulate-key to simulate the key presses?

zmughal commented:
That looks good. Much easier to use.

ChiragGhanshani committed:
Add test for keybindings

- Created attribute to access the scrolled window of the
  PageDrawingComponent.
- Increased the size of the Cairo::ImageSurface document so that the
  scrollbars work.
- Test Up, Down, Right, Left, PgUp, PgDown keys.

zmughal commented:
@ChiragGhanshani, you've really done weird things with the history on this branch! I need to undo the merging going on here.

zmughal commented:
It used to have 21 commits, but many of them were duplicates that happened because of merging. If your git client ever says that it has to merge when pulling, don't accept that. We need to keep a linear history on our feature branches.

Also, I added a commit to use the "cuddled" else style which is common in C and Perl code:

``` perl
if( ... ) {
  ...
} else { 
  ...
}
```

not

``` perl
if( ... ) {
  ...
}
else {
  ...
}
```

Consistency in style across a project is very important.

I want to fixup the commit https://github.com/project-renard/curie/pull/52/commits/23831be25e3bcf93c491e486188e24e4d46ef08e into https://github.com/project-renard/curie/pull/52/commits/14755ad6e15cd5ccfa9f9be28fb949b481f4a1eb before merging.

zmughal commented:
Also, @ChiragGhanshani, notice the commit message I edited into https://github.com/project-renard/curie/pull/52/commits/14755ad6e15cd5ccfa9f9be28fb949b481f4a1eb.

I took all the series of commits and put them in one so that all the keybindings get put together rather than incrementally changing them between different options. I also gave full sentence descriptions of what was done in a list format.

zmughal commented:
@ChiragGhanshani, what's left on this PR? Testing the arrow keys?

There may be a problem with that. We need to increase the size of the Cairo::ImageSurface's used to set up the test.

ChiragGhanshani commented:
@zmughal, yes. Just the arrow keys are left. I think I can test the up and down keys pretty easily, but the left and right ones will need zooming implemented.

StayerX commented:
@ChiragGhanshani you could split this pull request into two pull requests if you wish.One for the up and down and one for the left right key bindings. Either way there are other key bindings that will have t be implemented in the future if it makes sense.

zmughal commented:
I agree with @StayerX. We can leave off the left/right arrow tests until zooming is implemented after Clutter in #27.

zmughal commented:
By the way, ignore the commit history order on this PR page. GitHub's model is broken. It doesn't use the the Git DAG and assumes that you can always sort by date. But when you rebase and switch the commit order, that's not always true!

StayerX commented:
@zmughal you should create a pull request to fix it :P

zmughal commented:
Now that #70 is merged, you will need to rebase this on master and fix some merge conflicts. Also, `build_test` function can now be removed and use the function in the `CurieTestHelper` module.

zmughal commented:
@ChiragGhanshani, have time to rebase this off master and  do the rest of https://github.com/project-renard/curie/pull/52#issuecomment-207836004?

zmughal commented:
OK, rebased off of `master` branch. To get the changes, you can do:

``` shell
git fetch origin
git checkout curie-keybindings
git reset --hard origin/curie-keybindings
```

zmughal commented on t/Renard/Curie/Component/KeyBindings.t:
```diff
@@ -0,0 +1,66 @@
+use Test::Most;
+use Test::More;
```
FYI:

<del>`Test::Most` pulls in `Test::More` and a bunch of other useful testing modules. That way all the imports get out of your way and you only need to import `Test::Most`.</del> DONE

zmughal commented on t/Renard/Curie/Component/KeyBindings.t:
```diff
@@ -0,0 +1,66 @@
+  filename => CurieTestHelper->test_data_directory->child(qw(PDF Adobe pdf_reference_1-7.pdf))
+);
+
+sub build_test{
```
<del>I took the `build_test` code and put it into `CurieTestHelper` as `run_app_with_document`. You can use that instead.</del> DONE

zmughal commented on t/Renard/Curie/Component/KeyBindings.t:
```diff
@@ -0,0 +1,66 @@
+use Test::Most;
+use Test::More;
+
+use lib 't/lib/';
```
Nitpicky, but this would be better without the trailing slash:

``` diff
-use lib 't/lib/'; 
+use lib 't/lib'; 
```

zmughal replied:
DONE

zmughal commented on lib/Renard/Curie/App.pm:
```diff
@@ -46,6 +46,43 @@ sub setup_window {
+		my $current = $self->page_document_component->scrolled_window->get_vadjustment();
+		my $adjustment = $current->get_value() - $current->get_step_increment();
+		$current->set_value($adjustment);
+		$self->page_document_component->scrolled_window->set_vadjustment($current);
```
``` diff
+       $self->page_document_component->scrolled_window->set_vadjustment($current); 
```

This line (and the others below) are not necessary. The GtkAdjustment object is a reference, not a value.

zmughal replied:
DONE

zmughal commented on lib/Renard/Curie/Component/PageDrawingArea.pm:
```diff
@@ -118,6 +136,7 @@ sub setup_drawing_area {
 
 	$scrolled_window->add($drawing_area);
 	$scrolled_window->set_policy( 'automatic', 'automatic');
+  $self->scrolled_window($scrolled_window);
```
Fix the indent.

zmughal replied:
DONE

zmughal commented on t/Renard/Curie/Component/KeyBindings.t:
```diff
@@ -0,0 +1,66 @@
+});
+
+subtest 'Check that up arrow scrolls up and down arrow scrolls down' =>
+  build_test ( sub {
```
Fix the indent.

zmughal replied:
DONE

zmughal commented:
With some debugging, I found that the call back is sent and the code attempts to set the value of the GtkAdjustment, but under the test harness, I see

``` perl
$adj->set_value(50);
say "Value is: ", $adj->get_value;
# Value is: 0
```

I think that something is not initialising the scrolled window correctly to where it has the right bounds.

zmughal commented:
Made some fixes to make the tests run. Do this to continue work:

``` shell
git fetch origin
git reset --hard origin/curie-keybindings
```

zmughal commented on lib/Renard/Curie/App.pm:
```diff
@@ -46,6 +46,42 @@ sub setup_window {
+	} elsif($event->keyval == Gtk3::Gdk::KEY_Up){
+		my $current = $self->page_document_component->scrolled_window->get_vadjustment;
+		my $adjustment = $current->get_value - $current->get_step_increment;
+		$current->set_value($adjustment);
```
This code could be better as a method in the PageDrawingArea object.

e.g., `scroll_up_increment`, `scroll_down_increment`, `scroll_right_increment`, `scroll_left_increment`

zmughal replied:
Make these methods?

zmughal commented:
It's odd that the tests for this don't pass on Travis-CI, but run fine on Windows. Maybe we need to set up an Ubuntu Vagrant to get the same environment. It may have to do with the Gtk version.

zmughal commented:
It looks like page up and page down on Mac are [mapped](http://apple.stackexchange.com/questions/3687/are-there-keyboard-shortcuts-for-home-end-pagedown-and-pageup-in-os-x) to:
- fn + ↑ is PageUp
- fn + ↓ is PageDown
- fn + ← is Home
- fn + → is End

zmughal commented:
Odd that the code is failing on Travis-CI, but not on Appveyor. This might be because we're dealing with an old version of Gtk. Maybe running Vagrant is solution https://github.com/project-renard/devops/issues/14.

zmughal commented:
@ChiragGhanshani 
- make it pass on appveyor
- Ignore the Travis-CI failures, because that might be an old version of Gtk
- I will test on a local Linux machine
- We'll handle the Mac OS X issues in another PR

zmughal commented on t/Renard/Curie/Component/KeyBindings.t:
```diff
@@ -0,0 +1,45 @@
+
+my $cairo_doc = CurieTestHelper->create_cairo_document;
+
+sub Key_Event{
```
Can you stick this in CurieTestHelper?

ChiragGhanshani commented:
@zmughal, the tests are all passing on my machine. I'm not sure why the build is failing on the CI.

ChiragGhanshani commented:
Here's the output when I run prove.

[testOutput.txt](https://github.com/project-renard/curie/files/288145/testOutput.txt)

ChiragGhanshani commented:
Hmm, this one is a duplicate of an older pull request. Is there a way to merge the issues?

zmughal commented:
@ChiragGhanshani, which older PR? You mean the issue https://github.com/project-renard/curie/issues/31?

There is a way to merge them with Waffle.io. What you do is prefix your branch name with the issue number that it is fixing.

Check out:
- https://github.com/waffleio/waffle.io/wiki/FAQs#branch-moving
- https://github.com/waffleio/waffle.io/wiki/FAQs#prs-and-issues

As you can see on my branches, I've been doing `<issue-number>-the-branch-name`. See how they show up next to the issue they address https://waffle.io/project-renard/roadmap?

![firefox_2016-05-29_07-52-04](https://cloud.githubusercontent.com/assets/94489/15633437/3fb03d08-2572-11e6-941d-1149bddf4d5a.png)

zmughal commented:
It's passing here as well. I'm going to go ahead and rebase off master so that the conflicts get resolved.

zmughal commented on lib/Renard/Curie/App.pm:
```diff
@@ -57,24 +57,28 @@ sub key_pressed {
 	}
 }
 
+sub increment_scroll{
```
add a space between function name and `{` :-)

zmughal commented on lib/Renard/Curie/App.pm:
```diff
@@ -57,24 +57,28 @@ sub key_pressed {
+	$current->set_value($adjustment);
+}
+
+sub decrement_scroll{
```
add a space between function name and `{`

zmughal commented:
So I ran the whole application under the Vagrant box `ubuntu/precise64`. Oh my. It doesn't even display inside the window. I think the Glade file requires a newer Gtk?

I just ran it under `ubuntu/trusty64` and it works fine. These keybinding tests pass there. There is a way to use trusty on Travis-CI, so I will make an issue for that.

zmughal commented:
@ChiragGhanshani, please rebase this off master. Now that #110 has been closed, the tests should work properly on Travis-CI.

zmughal commented:
I'm going to go ahead and do the rebase and fixup. The failure is because I changed the size of the Cairo test document in this branch.

Is that OK, @ChiragGhanshani?

zmughal commented:
@ChiragGhanshani, ready to merge. Go for it.
